argon2/src/block.rs

```diff
@@ -67,6 +71,55 @@ impl Block {
     }
 
     pub(crate) fn compress(rhs: &Self, lhs: &Self) -> Self {
+        #[cfg(any(target_arch = "x86_64"))]
+        {
+            let (_, avx2) = avx2_cpuid::init_get();
+            if avx2 {
```

Comment:
This is calling `CPUID` for every invocation of `compress`. It'd be a lot better to find somewhere to store the init token (possibly in the `Argon2` struct, moving the `compress` function there, or at least the decision to use AVX2 or not).

Failing that you don't need to fetch and discard it:

```suggestion
            if avx2_cpuid::get() {
```

Reply:
`init_get` does not call CPUID every time, it properly caches results from the first call. ZST token acts as an initialization prove, which allows us to remove the third branch used for inner static initialization.

Reply:
They are not needed. `cpufeatures` already efficiently handles it.

Reply:
Ah, I see. So does that mean this is fine as is?

Reply:
So does that mean that this is fine as is? Or do we still need to keep the init token around.

Reply:
> ZST token acts as an initialization proof

...but it's being thrown away here

Reply:
>but it's being thrown away here

Yes, but lack of initialization proof does not mean that we disable caching of CPUID results. `avx2_cpuid::(init_)get()` simply has to [account](https://github.com/RustCrypto/utils/blob/master/cpufeatures/src/lib.rs#L149-L166) for possibility that the inner static could be uninitialized. While `token.get()` can [assume](https://github.com/RustCrypto/utils/blob/master/cpufeatures/src/lib.rs#L137-L143) that the static was already initialized (thus instead of 3 branches, the code will have only 2).